### PR TITLE
Add an `Iterator.prototype.toArray` to the adventure pack

### LINKED ISSUE #76: Add `Iterator.prototype.toArray` to the adventure pack
Not strictly necessary since there's spread syntax and `Array.from`, but being able to chain the method will be nice.

See also: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Iterator/toArray

### PR DESCRIPTION
**Pull request by elimanzo**
from: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Iterator/toArray

closes #76

### CODE REVIEW

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    expect([].values().toArray()).toEqual([]);
+  });
+
+  it("can chain with other iterator methods", () => {
+    const iterator = [1, 2, 3, 4, 5].values();
+    expect(
+      iterator
+        .toArray()
+        .map((x) => x ** 2)
+        .filter((x) => x % 2 === 0),
```
Comment (miorel) [inline tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts:21]: I like this test, but after the change `toArray`, these are no longer iterator methods, they are array methods. Is there another way to do this test?

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import "../Iterator.prototype.toArray";
+import { describe, expect, it } from "@jest/globals";
+
+describe("Iterator.prototype.toArray", () => {
+  it("can convert an iterator to an array", () => {
```
Comment (miorel) [inline tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts:5]:
```suggestion
  it("converts an iterator to an array", () => {
```

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    expect(
+      [1, 1, "a", 2, "b", "b", 3, "c"][Symbol.iterator]().toArray(),
+    ).toEqual([1, 1, "a", 2, "b", "b", 3, "c"]);
+  });
+
+  it("can return an empty array for an empty iterator", () => {
```
Comment (miorel) [inline tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts:11]:
```suggestion
  it("returns an empty array for an empty iterator", () => {
```

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import "../Iterator.prototype.toArray";
+import { describe, expect, it } from "@jest/globals";
```
Comment (miorel) [inline tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts:2]: Eventually I think we will need to add this as a lint rule, but I think a good style for imports is first external imports and then internal imports, separated by an empty line.

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    toArray(): T[];
+  }
+}
+
+iteratorPrototype.toArray ??= function <T>(this: Iterator<T>): T[] {
+  const res = [];
+  for (const element of this.toIterable()) {
+    res.push(element);
+  }
+  return res;
```
Comment (miorel) [inline tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.ts:19]: Does this work?

```suggestion
  return [...this.toIterable()];`
```

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import "../Iterator.prototype.toArray";
```
Comment (miorel) [inline tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts:1]: I think in the other tests I've been trying to write it as `import "./index";`

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts (hunk 2 above)
Comment (elimanzo) [reply tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts]: Ack!

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts (hunk 4 above)
Comment (elimanzo) [reply tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts]: Ack!

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.ts (hunk 5 above)
Comment (elimanzo) [reply tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.ts]: It does work!

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts (hunk 6 above)
Comment (elimanzo) [reply tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts]: Ack!

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts (hunk 3 above)
Comment (elimanzo) [reply tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts]: Ack!

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts (hunk 1 above)
Comment (elimanzo) [reply tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts]: Ack!

File: tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import { describe, expect, it } from "@jest/globals";
+
+import "./index";
+
+describe("Iterator.prototype.toArray", () => {
```
Comment (miorel) [inline tools/adventure-pack/goodies/typescript/Iterator.prototype.toArray/index.test.ts:5]: I think there are a couple more tests that could be interesting:

1. Write a generator function that throws an error, what happens if we try to call `.toArray()` on the resulting iterator? I'm thinking it ought to throw.
2. Test on an iterator that doesn't come from an array, e.g. a Set's `.values()`.

...but these can come as a follow-up PR!

**Review by miorel — APPROVED** [review]
Thanks for the PR!

**Pull request merged**